source/common/network/connection_impl.cc

```diff
@@ -553,7 +560,11 @@ void ConnectionImpl::onFileEvent(uint32_t events) {
   ScopeTrackerScopeState scope(this, this->dispatcher_);
   ENVOY_CONN_LOG(trace, "socket event: {}", *this, events);
 
-  if (immediate_error_event_ != ConnectionEvent::Connected) {
+  if (!immediate_error_event_.has_value()) {
```

Comment:
question: why are write events generated before connect is called? Seems a bit weird to have an event that denotes that we can write when we haven't even connected.

Reply:
Thanks. I add the my user space event here and it could active write an event before connected. Yeah, I agree I should also fix there so that event follows the convention. 

However, as a connection itself, the connection does register both write and read https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L93-L95

I think the best approach should be applied both at the connection and event system.

Connection as an event consumer should allow false activation, and event producer should produce as few false alarm as possible

Reply:
> However, as a connection itself, the connection does register both write and read

we need to register `write` because we use it to communicate bind/connect failures. This is consistent with the fact that the kernel will send a `write` when the connection is established in the happy path.

After you call `connect`, how do you know that the first `write` event that you get indicates that connection is ready vs a 'false' event?

Reply:
basically the following sequence:

1. Call `connect` -> `SOCKET_ERROR_IN_PROGRESS` -> `immediate_error_event_ = ConnectionEvent::Connected;`
2. You get a `Write` ready event.

If this is a false event then couldn't the following call fail
```
    RELEASE_ASSERT(
        socket_->getSocketOption(SOL_SOCKET, SO_ERROR, &error, &error_size).return_value_ == 0, "");
```
But it could succeed on the next `write` event which is the one that indicates that the connection is ready

Reply:
> After you call `connect`, how do you know that the first `write` event that you get indicates that connection is ready vs a 'false' event?

I didn't change the write/read event registration in this PR. This PR is restricted to when to set immediate_error_event_

See the new connect() in this PR. I set the immediate_event_ there.

So after connect() is called, W/R is registered as in the past and immediate_error_event_ is set to either localclose(on error) or connected(as was in the past)

Reply:
Fair enough, this is not related to the patch.

I am just wondering if with this PR we address all the issues with false write events or they expose more problems

Reply:
@davinci26 Right. There is definitely some risk here 

Could you join the discussion ? https://github.com/envoyproxy/envoy/issues/18979#issuecomment-967331859